gpii/node_modules/lifecycleManager/src/LifecycleManager.js

```diff
@@ -46,20 +46,36 @@ var gpii = fluid.registerNamespace("gpii");
                     gpiiKey: "{arguments}.1"
                 },
                 createOnEvent: "onSessionStart"
+            },
+            userLogonRequests: {
+                type: "{arguments}.0",
+                createOnEvent: "onUserLogonRequestReceived",
+                options: {
+                    gpiiKey: "{arguments}.1",
+                    requestPromise: "{arguments}.2"
+                }
             }
         },
         members: {
             sessionIndex: {}, // map of gpiiKey to session component member name, managed by gpii.lifecycleManager.sessionIndexer
-            /* queue for high-level lifecycle manager tasks (start, stop or update).
-             * The entries in the queue are of the format { func: <functionToCall>, arg: <argument> } where
+            /* action queue for high-level lifecycle manager tasks (start, stop or update).
+             * The entries in the action queue are of the format { func: <functionToCall>, arg: <argument> } where
              * <functionToCall> is a single-argument function that returns a promise. The promise should be resolved when
              * the function is complete (including side-effects).
              * A more detailed description: https://github.com/GPII/universal/tree/master/documentation/LifecycleMananger.md
              */
-            queue: [],
-            isProcessingQueue: false // if queue is currently being processed
+            actionQueue: [],
+            userLogonRequestQueue: [],
+            isProcessingActionQueue: false, // if action queue is currently being processed
```

Comment:
See if we can do without these flags - every bit of shared mutable state carries an extra risk of races. Is it not enough to check whether the length of the queue is 0?

Reply:
Tests show these flags are still required. Since the processing of a user logon request or a action is asynchronous, this flag is important to indicates the period from when a queue item has been removed from the queue to process to the end of the processing. 

Here's [an example](https://github.com/cindyli/universal/blob/GPII-3193/gpii/node_modules/lifecycleManager/src/UserLogonRequest.js#L219): Determining whether "noUser" keyin should be added to the queue checks both the length of the queue is 0 and the `isProcessingUserLogonRequestQueue` flag. This is to cover the case when the last request on the queue, for example an actual user keyin, has been removed from the queue (the length of the queue is now 0), but is still in process (`isProcessingUserLogonRequestQueue === true`), "noUser" keyin would not be added to the queue.

Reply:
Couldn't we resolve this by only removing an item from the queue when it has finished processing, rather than when it starts? This would have the additional benefit of having an obvious public place to look in order to determine which item is currently being processed